[iOS - Photos 라이브러리] collectionView 생성

1Consumption opened the pull request:
• 셀사이즈를 80 * 80으로 고정하기 위해서 UICollectionViewDelegateFlowLayout를 채택했습니다.
• viewController부터 view class와 model class를 분리했습니다.
• random한 color 값을 생성하기위해 generateRandomColor()를 구현하였고, alpha 값은 0.5 에서 1사이로 설정했습니다.
* feature를 기능 단위로 구분했습니다.

1Consumption committed:
[기능 추가] Readme 수정

원활한 협업을 위한 컨벤션 정의

1Consumption committed:
[기능 수정] SceneDelegate 삭제

iOS 13 미만 버전 호환성 고려

1Consumption committed:
[기능 추가] view 클래스 생성

viewController의 최상위 view객체를 생성

1Consumption committed:
[기능 추가] collectionView cell random color 적용

램덤컬러를 반환하는 generateRandomColor() 메소드 작성했습니다.
알파값이 0.5부터 주어진 이유는 너무 낮으면 색상 표현이 안돼서 0.5 이상으로 설정했습니다.

1Consumption committed:
[기능 추가] collectionView cell 크기 설정

UICollectionViewDelegateFlowLayout 프로토콜을 채택하여 80 * 80의 크기로 고정했습니다.

godrm commented on PhotosApp/PhotosApp/DataManager.swift:
```diff
@@ -0,0 +1,42 @@
+        
+        let alpha = CGFloat.random(in: (0.5..<1))
+        
+        return UIColor(red: red, green: green, blue: blue, alpha: alpha)
```
랜덤 컬러를 만드는 일이 여기 객체가 갖고 있는게 최선의 선택일까요?

GangWoon replied:
UICollectionView cell 클래스를 만들어서 처리하라는 말씀이신가요 ??

godrm replied:
지금은 셀이 꼭 필요한 상황은 아니지 않나요? 매개변수와 리턴 타입을 보고 유추해보세요.

godrm commented on PhotosApp/PhotosApp/DataManager.swift:
```diff
@@ -0,0 +1,42 @@
+
+extension DataManager: UICollectionViewDelegateFlowLayout {
+    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
+        return CGSize(width: 80, height: 80)
```
숫자나 문자열도 하드 코딩하지 말고 의미있는 코드로 표현해보세요. 
"collectionViewCell" Cell Identifier는 어느 객체와 관련이 있는 값일까요? 
다른 셀이 여러 개 있다고 생각해보세요.

godrm commented on PhotosApp/PhotosApp/PhotosView.swift:
```diff
@@ -0,0 +1,14 @@
+
+class PhotosView: UIView {
+    
+    @IBOutlet weak var collectionView: UICollectionView!
```
세부적인 기능이나 차이가 전혀 없다면 그냥 콜랙션뷰를 써도 될 것 같은데 왜 UIView를 만들었나요?

GangWoon replied:
view 클래스가 확장될 거 같아서 UIView를 상속 받았었습니다.
하지만 페어와 커뮤니케이션한 결과 지금 상황에서는 view 클래스를 나누는게 의미가 없어 보여서 다시 합치기로 결정했습니다.

godrm replied:
그럼 상속을 받는것과 포함관계를 갖는 것이 어떤 차이가 있죠? 
이런 구조라면 self.view 에 collectionView를 add해도 동일한 거 아닌가요?

godrm commented on PhotosApp/PhotosApp/ViewController.swift:
```diff
@@ -10,11 +10,17 @@ import UIKit
-        // Do any additional setup after loading the view.
+        
+        photosView.collectionView.dataSource = dataManager
+        photosView.collectionView.delegate = dataManager
```
심지어 이렇게 내부 뷰에 접근할꺼면 그냥 Self.view를 CollectionView로 만들어도 되는게 아닐까요? 
지금 델리게이트까지 필요한가요?

GangWoon replied:
델리게이트 속성을 datamanager에게 넘긴 이유는 셀 크기를 변경하기 위해서 작성했습니다.

godrm replied:
그랬군요. 그럼 dataSource와 delegate가 같은 객체여야만 하는건가요?

GangWoon replied:
처리했습니다~

godrm replied:
어떻게 개선했는지 댓글에도 작성해주면 리뷰하는 데 도움이 될 것 같습니다.

GangWoon replied:
datasource는 모델 객체에서 여전히 처리하는게 맞다고 생각했습니다.
하지만 delegate는 모델 객체가 아닌 controller에서 처리하는게 맞다고 방금 생각이 들었습니다.(원래는 view에서 처리한다고 생각했습니다.)
갑자기 생각을 바꾼 이유는 셀을 클릭했을 때 다음 화면으로 넘어가는 상황을 생각했고, 이런 경우에 view 클래스와 model 클래스에서 controller를 알고 있는게 부자연스럽다고 생각이 되어 controller에서 처리하는게 맞다고 생각했습니다!!

godrm reviewed (changes requested):
다음 내용을 수정하고 확인되면 머지하셔도 됩니다.

GangWoon committed:
[기능 수정] 분리했던 view 클래스를 병합

서로 커뮤니케이션을 진행한 결과 아직까지 뷰 클래스를 나눌 이유를 찾지 못해서 다시 합쳤습니다.

GangWoon committed:
[기능 추가] CollectionViewDatasource 클래스 생성

collectionView의 데이터를 처리할 클래스를 생성했습니다.
delegate의 처리도 같은 객체에서 처리했지만, delegate는 모델 클래스가아닌 뷰 클래스에서 처리할는게  더 적합해 보였습니다.

GangWoon committed:
[기능 수정] UICollectionViewCell 클래스 랜덤 컬러 생성

셀에서 자신의 백그라운드 컬러를 변경할 수 있도록 수정했습니다.

GangWoon committed:
[기능 수정]  delegate 객체 삭제

의견을 나눈 결과 collectionView.delegate는 view 객체에서 처리하는게 맞다고 생각해서 viewcontroller에서 처리했습니다.

GangWoon committed:
[폴더 생성 커밋]

godrm commented on PhotosApp/PhotosApp/Extension/UICollectionViewCell.swift:
```diff
@@ -0,0 +1,20 @@
+        let green: CGFloat = CGFloat.random(in: range) / 255
+        let blue: CGFloat = CGFloat.random(in: range) / 255
+        let alpha: CGFloat = CGFloat.random(in: (0.5..<1))
+        backgroundColor = UIColor(red: red, green: green, blue: blue, alpha: alpha)
```
저는 UIColor를 다루는 객체니까 UIColor를 확장하는 게 어떨까 생각했습니다. 
모든 셀이 필요한 기능은 아니지 않을까요?

godrm commented on PhotosApp/PhotosApp/Model/CollectionViewDataSource.swift:
```diff
@@ -0,0 +1,33 @@
+class CollectionViewDataSource: NSObject {
+    
+    private let count: Int
+    private let cellIdentifier: String = "photoCell"
```
cellIdentifier 이 값이야 말고 UICollectionViewCell를 관련 값인 것 같네요. 셀마다 다른 값이니까요.

godrm commented:
머지 처리하시고 다음 스텝을 진행하세요.

GangWoon committed:
[기능 추가] randomColor 생성

UIcollectonViewCell에서 처리하는게 아닌, UIColor에서 처리하도록 구현했습니다.

GangWoon committed:
[기능 추가] photoCellIdentifier 추가

셀마다 다른 identifier를 갖을 수 있어서 하드 코딩을 지양하고자 static 상수로 선언했습니다.

GangWoon committed:
[기능 수정] cell 설정부 변경

cell에서 값을 만들어서 사용하는게 아닌 uicolor에서 랜덤 컬러를 반환하도록 하여 셀에서는 백그라운드 컬러 값만 변경하도록 했습니다.
cell identifier를 하드코딩하지 않기위해서 상수화 했습니다.

GangWoon committed:
[폴더 생성 커밋]
